tutorials/tour/_posts/2017-02-13-extractor-objects.md

```diff
@@ -11,25 +11,25 @@ next-page: sequence-comprehensions
 previous-page: regular-expression-patterns
 ---
 
-In Scala, patterns can be defined independently of case classes. To this end, a method named unapply is defined to yield a so-called extractor. An extractor can be thought of as a special method that reverses the effect of applying a particular object on some inputs. Its purpose is to 'extract' the inputs that were present before the 'apply' operation. For instance, the following code defines an extractor [object](singleton-objects.html) Twice.
-
-```tut
-object Twice {
-  def apply(x: Int): Int = x * 2
-  def unapply(z: Int): Option[Int] = if (z%2 == 0) Some(z/2) else None
-}
-
-object TwiceTest extends App {
-  val x = Twice(21)
-  x match { case Twice(n) => Console.println(n) } // prints 21
-}
+An extractor object is an object with an `unapply` method. Whereas the a `apply` method is like a constructor which takes arguments and creates an object, the `unapply` takes an object and tries to give back the arguments.
 ```
+import scala.util.Random
 
-There are two syntactic conventions at work here:
+object CustomerID{
 
-The pattern `case Twice(n)` will cause an invocation of `Twice.unapply`, which is used to match any even number; the return value of the `unapply` signals whether the argument has matched or not, and any sub-values that can be used for further matching. Here, the sub-value is `z/2`
+  def apply(name: String) = s"$name--${Random.nextLong}"
 
-The `apply` method is not necessary for pattern matching.  It is only used to mimick a constructor. `val x = Twice(21)` expands to `val x = Twice.apply(21)`.
+  def unapply(customerID: String): Option[String] = {
+    val name = customerID.split("--")(0)
+    if (!name.isEmpty) Some(name) else None
```

Comment:
You could use `if (name.nonEmpty) …`.

On the line above, you could also use `customerID.split("--").head`, or even `.headOption`, but `split` always returns at least one element, so I think that’s fine…